### PR TITLE
[qt6] Use RecursiveMutex instead of QMutex( QMutex::Recursive )

### PR DESCRIPTION
**Pull request by m-kuhn**
QMutex::Recursive will be gone.
The replacement only entered the stage with Qt 5.14

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -438,7 +438,7 @@ if(WITH_CORE)
     set(QT_MIN_VERSION 6.0.0)
     set(QT_VERSION_BASE "Qt6")
   else()
-    set(QT_MIN_VERSION 5.12.0)
```
Comment (nyalldawson) [inline CMakeLists.txt:441]: Mmmm.... We had good reasons to limit this to 5.12... I think we'll need to just ifdef in the replacement here

**Review by nyalldawson — APPROVED** [review]

File: src/core/auth/qgsauthmethod.h [hunk 2]
```diff
@@ -173,7 +178,6 @@ class CORE_EXPORT QgsAuthMethod : public QObject
      * \note Non-public since this is an abstract base class
      */
     explicit QgsAuthMethod()
-      : mMutex( QMutex::RecursionMode::Recursive )
```
Comment (nyalldawson) [inline src/core/auth/qgsauthmethod.h:176]: #if QT_VERSION < QT_VERSION_CHECK(5, 14, 0)
    : mMutex( QMutex::RecursionMode::Recursive )
#endif

(github won't let me suggest this!)

File: src/core/providers/ogr/qgsogrprovider.h [hunk 3]
```diff
@@ -397,13 +397,17 @@ class CORE_EXPORT QgsOgrProviderUtils
...
+#else
+        QRecursiveMutex mutex;
+#endif
+        GDALDatasetH    hDS = nullptr;
         QMap<QString, QgsOgrLayer *>  setLayers;
         int            refCount = 0;
         bool           canBeShared = true;
 
-        DatasetWithLayers(): mutex( QMutex::Recursive ) {}
+        DatasetWithLayers() {}
```
Comment (nyalldawson) [inline src/core/providers/ogr/qgsogrprovider.h:410]: DatasetWithLayers()
#if QT_VERSION < QT_VERSION_CHECK(5, 14, 0)
: mutex( QMutex::Recursive )
#endif 
{}

File: src/core/auth/qgsauthmethod.h [hunk 4]
```diff
@@ -173,7 +178,9 @@ class CORE_EXPORT QgsAuthMethod : public QObject
      * \note Non-public since this is an abstract base class
      */
     explicit QgsAuthMethod()
+#if QT_VERSION < QT_VERSION_CHECK(5, 14, 0)
       : mMutex( QMutex::RecursionMode::Recursive )
+#endif
     {}
```
Comment (m-kuhn) [inline src/core/auth/qgsauthmethod.h:184]:
```suggestion
#ifndef SIP_RUN
#if QT_VERSION < QT_VERSION_CHECK(5, 14, 0)
      : mMutex( QMutex::RecursionMode::Recursive )
#endif
    {}
#endif
```

File: python/core/auto_generated/auth/qgsauthmethod.sip.in [hunk 5]
```diff
@@ -152,6 +153,8 @@ Construct a default authentication method
 
    Non-public since this is an abstract base class
 %End
+      : mMutex( QMutex::RecursionMode::Recursive );
+    {}
```
Comment (m-kuhn) [inline python/core/auto_generated/auth/qgsauthmethod.sip.in:157]:
```suggestion
```

**Pull request merged**